keps/sig-node/2000-graceful-node-shutdown/README.md

```diff
@@ -0,0 +1,864 @@
+      running user workloads?
+    - Diagnostics: What are the useful log messages and their required logging
+      levels that could help debug the issue?
+      Not required until feature graduated to beta.
+    - Testing: Are there any tests for failure mode? If not, describe why.
+
+* **What steps should be taken if SLOs are not being met to determine the problem?**
+
+[supported limits]: https://git.k8s.io/community//sig-scalability/configs-and-limits/thresholds.md
+[existing SLIs/SLOs]: https://git.k8s.io/community/sig-scalability/slos/slos.md#kubernetes-slisslos
+
+## Implementation History
+
+<!--
+Major milestones in the lifecycle of a KEP should be tracked in this section.
+Major milestones might include:
+- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
+- the `Proposal` section being merged, signaling agreement on a proposed design
+- the date implementation started
+- the first Kubernetes release where an initial version of the KEP was available
+- the version of Kubernetes where the KEP graduated to general availability
+- when the KEP was retired or superseded
+-->
+
+*   May 26 - [Original GH issue #91472
+    filed](https://github.com/kubernetes/kubernetes/issues/91472)
+
+## Drawbacks
+
+<!--
+Why should this KEP _not_ be implemented?
+-->
+
+## Alternatives
+
+<!--
+What other approaches did you consider, and why did you rule them out? These do
+not need to be as detailed as the proposal, but should include enough
+information to express the idea and why it was not acceptable.
+-->
+
+* Use systemd cgroup driver to set `TimeoutStopSec=` on scopes underlying
+  containers
+    * Set `TimeStopSec=` for the container scopes using the value set in the pod
+      for termination grace period. The problem with this approach is that
+      systemd doesn’t understand the prestop hooks. 
+* Use systemd cgroup driver to set `Before=kubelet.service` on scopes
+  underlying containers
+    * Set `Before=kubelet.service` and container runtime service for the
+      container scopes. Systemd would then stop the containers after the
+      kubelet giving the kubelet a chance to stop the containers itself. This
+      depends upon using the systemd cgroups driver and is coupled to systemd.
+* Use systemd cgroup driver to set controller property on scope to delegate
+  control to kubelet
+    * Set Controller dbus property for the container scopes and set
+      `After=kubelet.service` for the containers. Systemd would then signal the
+      kubelet over dbus to delegate the container scope termination. This
+      requires more work in the kubelet and is also coupled to systemd and the
+      systemd cgroup driver.
+* Don’t handle node shutdown events at all, and have users drain nodes before
+  shutting them down.
+    * This is not always possible, for example if the shutdown is controlled by
+      some external system (e.g. Preemptible VMs).
```

Comment:
Does "Preemptible VMs" in all cloud providers trigger the shutdown event on termination?

Reply:
(assuming the OS for the VM is running a compatible systemd version)

Reply:
That's a good question -- every cloud provider will eventually have to shutdown the VM and terminate it, so at some point the VM shutdown should be sent. 

For example, on GCE when a GCE Preemptible VM is terminated it gets 30 second time to shutdown, and the shutdown event is delivery at t-30 where t is when it will be forced shutdown. On AWS spot instances, the period is 2 minutes, but I'm unclear if the shutdown event is delivered a t-2min or at t itself. In addition to the systemd shutdown event, each cloud provider usually has a specific metadata server local to the VM that can be used to poll for preemption events. 

If the specific cloud provider doesn't actually trigger shutdown prior to the VM getting preempted, one workaround would be to have a cloud specific daemonset deployed that can poll the metadata server of the VM for preemption event and upon receiving the terminating event, simply trigger a node shutdown, i.e. `systemctl poweroff`, which will initiate kubelet graceful shutdown.